File: Reorder.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Y

/* 1. split the list from middle
    2. reverse the second list
    3. merge both list
*/
public class Reorder {
    
    public void reorderList(ListNode head) {
        
        if(head == null || head.next == null || head.next.next == null)
            return;

        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next !=null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }    
        fast = reverse(slow.next);
        slow.next = null;
        slow = head;
        while(fast!=null)
        {
            ListNode temp = slow.next;
            slow.next = fast;
            fast = fast.next;
            slow.next.next = temp;
            slow = temp;
        }
    }

    public ListNode reverse(ListNode head) {
        if(head == null || head.next == null)
            return head;
        ListNode prev = null;
        ListNode curr = head;
        while(curr!=null)
        {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
          
            curr = temp;
        }
       
        return prev;
    }

    public static void main(String args [])
    {
        Reorder list = new Reorder();
        ListNode node = new ListNode(1);
        node.next = new ListNode(2);
        node.next.next = new ListNode(3);
        node.next.next.next = new ListNode(4);
      
 
     
        list.printList(node);
        list.reorderList(node);
        System.out.println("After Reorder");
        list.printList(node);
    
    
    }

    void printList(ListNode node)
    {
        while (node != null) {
            System.out.println(node.val + " ");
            node = node.next;
        }
    }
}

File: reorder.cpp


// first finds the middle node of the list using the slow and fast pointer technique.
// splits the list into two halves at the middle node.
// reverses the second half of the list.
// merges the two halves by alternating nodes from each half.
// Overall, the code has a time complexity of O(n) and a space complexity of O(1), 
// where n is the length of the linked list.

class Solution {
    public void reorderList(ListNode head) {
        if(head==null)
            return;

        // Initialize fast and slow pointers to find the middle node
        ListNode fast,slow,temp;
        fast=head;
        slow=head;

        while(fast!=null && fast.next!=null)
        {
            fast=fast.next.next;
            slow=slow.next;
        }
    
        // Split the list into two halves
        fast=slow.next;
        slow.next=null;
        slow=fast;

        ListNode head2=null;
        
        // Reverse the second half of the list
        while(slow!=null)
        {
            slow=slow.next;
            fast.next=head2;
            head2=fast;
            fast=slow;
        }
        
        slow=head;
        temp=head;
        fast=head2;

        // Merge the two halves by alternating nodes from each half
        while(fast!=null)
        {
            slow=slow.next;
            temp.next=head2;
            fast=fast.next;
            head2.next=slow;
            temp=slow;
            head2=fast;
        }
    }
}
